### DISCUSSION TITLE
Running rubocop without bundle exec significantly reduces run time (w/ benchmarks)

### DISCUSSION THREAD

**Discussion by mculp** [evilmartians/lefthook#410] [Q&A] [answered]
I read in a rubocop PR that running rubocop (and erblint) without `bundle exec` speeds up boot time by ~300ms for the individual who made the post. So, I removed `bundle exec` from my rubocop and erblint lefthook configs and tried it, and it indeed reduce run time consistently and significantly.

### Config changes
<img width="895" alt="image" src="https://user-images.githubusercontent.com/129983/210496547-0d03668f-3aa1-440f-85d7-0362eb465549.png">

I noticed that after I did this, the `pre-commit` script changed:

<img width="893" alt="image" src="https://user-images.githubusercontent.com/129983/210495759-e79cad9c-8818-48ea-8caf-c59fe68b3376.png">

I guess I have two questions:
 
1) I'm wondering what exactly this is doing and why it changed after removing `bundle exec` from the config? I've noticed these files changing before, too. I don't have `pnpm` on my machine. The first section is just moved and I don't have `pnpm` on my machine. So it doesn't appear to actually change how anything is executed.

2) To make this safe for my team, who could be running a different setup, should I check that `rubocop` exists with `type rubocop` or something similar or is that something this script handles?

Generating a binstub and using `bin/rubocop` does not give the performance boost, either.

A note for anyone else interested, if you run `rubocop --start-server` or add `--server` to the rubocop command, it will start a rubocop server with its files already loaded and the second time you run rubocop, it can drop your runtime significantly more.

### `bundle exec rubocop --no-server ...`
<img width="378" alt="image" src="https://user-images.githubusercontent.com/129983/210498271-8b7dfff8-110c-42ab-b2aa-b1e50126ad3c.png">

### `rubocop --no-server ...`
<img width="387" alt="image" src="https://user-images.githubusercontent.com/129983/210499327-4fc05b9f-3681-4ca8-90c1-f38d49d43fe3.png">

### `bundle exec rubocop --server ...` (second run with server running)
<img width="376" alt="image" src="https://user-images.githubusercontent.com/129983/210499512-87fb11c1-2f18-442c-8afa-befd91b10ca3.png">

### `rubocop --server` (second run with server running)
<img width="446" alt="image" src="https://user-images.githubusercontent.com/129983/210497671-bbc8f3cf-4c65-4e9f-b43f-e4fea83e3ffb.png">

Note: I used `git restore .git-hooks/` to remove the `pre-commit` and `prepare-commit-msg` scripts between these first four. The updated scripts seem to drop it a bit more. Seems consistent ~0.26-0.27s without the scripts and ~0.18-0.19s with. Maybe the extra time is Lefthook generating the new scripts? I don't see anything from the scripts that would change anything.

### `rubocop --server` with updated scripts
<img width="365" alt="image" src="https://user-images.githubusercontent.com/129983/210500573-31db410e-7218-4241-8724-2f2c8a73798e.png">

**Comment by mrexox** [comment, marked as answer]
Hey! Very cool suggestion, I really like the boost using `rubocop --server`. This is really useful!

I want to answer a few questions:
1. Lefthook uses a hook template which finds the lefthook executable and runs it. Since lefthook is distributed as NPM package, Ruby Gem, simple binary, it can be missing in $PATH, so that's why we do `bundle exec`, `pnpm`, `npx`, and check node_modules. Adding `pnpm` was a new thing in last released version, and you might have updated lefthook recently, so this change appeared.
2. You can provide a [`fail_text`](https://github.com/evilmartians/lefthook/blob/master/docs/configuration.md#fail_text) with instructions to install rubocop globally :) Or you can create a more sophisticated solution providing not a command, but a [script](https://github.com/evilmartians/lefthook/blob/master/docs/configuration.md#script), which checks for rubocop and installs it is it is missing. This requires some "dances" with lefthook but IMHO providing a fail_text with s fixing command is just enough.
